File: front/apps/web-ele/src/api/core/role.ts
import { requestClient } from '#/api/request';

export interface Role {
  id: number;
  name: string;
  code: string;
  description: string;
  status: string;
  level: number;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface RoleListResponse {
  total: number;
  roles: Role[];
}

/**
 * 获取职位列表
 */
export async function getRoleListApi(params?: {
  status?: string;
  page?: number;
  page_size?: number;
}): Promise<RoleListResponse> {
  const response = await requestClient.get<RoleListResponse>('/role', { params });
  return response.data;
}

/**
 * 获取职位详情
 */
export async function getRoleDetailApi(id: number): Promise<Role> {
  const response = await requestClient.get<Role>(`/role/${id}`);
  return response.data;
}

/**
 * 创建职位
 */
export async function createRoleApi(data: {
  name: string;
  code: string;
  description?: string;
  level?: number;
}): Promise<Role> {
  const response = await requestClient.post<Role>('/role', data);
  return response.data;
}

/**
 * 更新职位
 */
export async function updateRoleApi(
  id: number,
  data: {
    name?: string;
    code?: string;
    description?: string;
    level?: number;
  },
): Promise<Role> {
  const response = await requestClient.put<Role>(`/role/${id}`, data);
  return response.data;
}

/**
 * 停用职位
 */
export async function deactivateRoleApi(id: number): Promise<{ message: string }> {
  const response = await requestClient.put<{ message: string }>(`/role/${id}/deactivate`);
  return response.data;
}

/**
 * 激活职位
 */
export async function activateRoleApi(id: number): Promise<{ message: string }> {
  const response = await requestClient.put<{ message: string }>(`/role/${id}/activate`);
  return response.data;
}

File: back/internal/user/domain/role.go
package domain

import (
	"time"
	"gorm.io/gorm"
)

// 职位状态常量
const (
	RoleStatusActive   = "active"   // 激活
	RoleStatusInactive = "inactive" // 停用
)

// Role 职位聚合根
type Role struct {
	ID          uint           `gorm:"primaryKey" json:"id"`
	Name        string         `gorm:"size:100;not null" json:"name"`
	Code        string         `gorm:"size:50;not null;uniqueIndex" json:"code"`
	Description string         `gorm:"size:500" json:"description"`
	Status      string         `gorm:"size:20;default:active" json:"status"`
	Level       int            `gorm:"default:0" json:"level"`
	CreatedAt   time.Time      `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt   time.Time      `gorm:"autoUpdateTime" json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName 表名
func (Role) TableName() string {
	return "roles"
}

// Validate 验证职位数据
func (r *Role) Validate() error {
	if r.Name == "" {
		return ErrRoleNameEmpty
	}

	if len(r.Name) < 2 || len(r.Name) > 100 {
		return ErrRoleNameInvalid
	}

	if r.Code == "" {
		return ErrRoleCodeEmpty
	}

	if len(r.Code) < 2 || len(r.Code) > 50 {
		return ErrRoleCodeInvalid
	}

	return nil
}

// Deactivate 停用职位
func (r *Role) Deactivate() {
	r.Status = RoleStatusInactive
}

// Activate 激活职位
func (r *Role) Activate() {
	r.Status = RoleStatusActive
}

// RoleRepository 职位仓储接口
type RoleRepository interface {
	Create(role *Role) error
	Update(role *Role) error
	FindByID(id uint) (*Role, error)
	FindByCode(code string) (*Role, error)
	List(status *string, page, pageSize int) ([]*Role, int64, error)
	Delete(id uint) error
}
